### PR TITLE
add support for ssl options

### PR DESCRIPTION
**Pull request by choonkeat**
update `ConfigStringToClusterConfig` to understand a few more url params

- `enableHostVerification=true`
- `certPath`
- `keyPath`
- `caPath`

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -154,12 +155,33 @@ func ConfigStringToClusterConfig(configString string) (*gocql.ClusterConfig, err
 				case "password":
 					passwordAuthenticator.Password = value
 					clusterConfig.Authenticator = passwordAuthenticator
+				case "enableHostVerification":
+					if sslOptions == nil {
+						sslOptions = &gocql.SslOptions{}
+					}
+					sslOptions.EnableHostVerification = (value == "true")
```
Comment (MichaelS11) [inline config.go]: What do you think about using https://golang.org/pkg/strconv/#ParseBool ?

File: config.go [hunk 2]
```diff
@@ -84,6 +84,7 @@ func ConfigStringToClusterConfig(configString string) (*gocql.ClusterConfig, err
 	}
 
 	passwordAuthenticator := gocql.PasswordAuthenticator{}
+	var sslOptions *gocql.SslOptions
```
Comment (MichaelS11) [inline config.go]: What happens if sslOptions is set to `&gocql.SslOptions{}` and all the if `sslOptions == nil {` are removed? Is an empty `&gocql.SslOptions{}` the same as a nil one?

If it is the same, then can just set `&gocql.SslOptions{}` in clusterConfig and remove `var sslOptions`

File: config_test.go [hunk 3]
```diff
@@ -74,6 +89,7 @@ func TestConfigStringToClusterConfig(t *testing.T) {
 		{info: "invalid key", configString: "?foo=bar", err: fmt.Errorf("invalid key: foo")},
 
 		{info: "empty", configString: "", clusterConfig: NewClusterConfig()},
+		{info: "empty", configString: sslConfigString, clusterConfig: sslClusterConfig},
```
Comment (MichaelS11) [inline config_test.go]: Please add some more tests, including testing for empty data and bool test.

File: config.go [hunk 4]
```diff
@@ -154,12 +155,33 @@ func ConfigStringToClusterConfig(configString string) (*gocql.ClusterConfig, err
...
+				case "enableHostVerification":
+					if sslOptions == nil {
+						sslOptions = &gocql.SslOptions{}
+					}
+					sslOptions.EnableHostVerification = (value == "true")
+				case "certPath":
+					if sslOptions == nil {
+						sslOptions = &gocql.SslOptions{}
+					}
+					sslOptions.CertPath = value
```
Comment (MichaelS11) [inline config.go]: For this one and the next two, was just wondering is there any special checks that should be done? If not, no worries. Just was curious.

**Review by MichaelS11 — CHANGES_REQUESTED** [review]
Thank you very much for this PR! :)

File: config.go (hunk 1 above)
Comment (choonkeat) [reply config.go]: oops, will do

File: config.go (hunk 2 above)
Comment (choonkeat) [reply config.go]: [they are treated the same](https://github.com/gocql/gocql/blob/95d072f1b5bbc604ce10f41ce1c914578592ed00/connectionpool.go#L30-L33). is it robust to rely on that?

Comment (MichaelS11) [reply config.go]: Ok then, but is var sslOptions needed? Can clusterConfig.SslOpts be used directly?

File: config.go (hunk 4 above)
Comment (choonkeat) [reply config.go]: i think it's better for this function to just convert the `string` to `*gocql.ClusterConfig` and leave any further validation elsewhere

File: config.go (hunk 2 above)
Comment (choonkeat) [reply config.go]: > they are treated the same. is it robust to rely on that?

oops i'd referenced the wrong code last night. it's actually [here](https://github.com/gocql/gocql/blob/95d072f1b5bbc604ce10f41ce1c914578592ed00/connectionpool.go#L85-L86)

``` go
if cfg.SslOpts != nil {
    tlsConfig, err = setupTLSConfig(cfg.SslOpts)
```

so actually an empty `&gocql.SslOptions{}` isn't treated the same as a `nil` one

---

> Ok then, but is var sslOptions needed? Can clusterConfig.SslOpts be used directly?

we could use `clusterConfig.SslOpts` directly with `if ...` or I think we should follow how you approached `clusterConfig.Authenticator`

``` go
case "username":
	passwordAuthenticator.Username = value
	clusterConfig.Authenticator = passwordAuthenticator
case "password":
	passwordAuthenticator.Password = value
	clusterConfig.Authenticator = passwordAuthenticator
```

and do this

``` go
case "certPath":
	sslOpts.CertPath = value
	clusterConfig.SslOpts = &sslOpts
case "keyPath":
	sslOpts.KeyPath = value
	clusterConfig.SslOpts = &sslOpts
case "caPath":
	sslOpts.CaPath = value
	clusterConfig.SslOpts = &sslOpts
```

File: config_test.go (hunk 3 above)
Comment (choonkeat) [reply config_test.go]: as I added some tests, I found there's a bug https://github.com/MichaelS11/go-cql-driver/blob/60b83bcdb2ef5c698b6481351b2508f3c1734a93/config.go#L96

1. we actually shouldn't be doing `strings.TrimSpace` for the value
1. we should `url.QueryUnescape` the value, esp since `password=` could be a string with special characters

``` go
key, rawValue := strings.TrimSpace(settingSplit[0]), settingSplit[1]
value, err := url.QueryUnescape(rawValue)
if err != nil {
	return nil, fmt.Errorf("failed for: %v = %v", key, rawValue)
}
```

should I add this into this PR too?

Comment (choonkeat) [reply config_test.go]: 7d8f510

File: config.go [hunk 5]
```diff
@@ -93,7 +95,11 @@ func ConfigStringToClusterConfig(configString string) (*gocql.ClusterConfig, err
 				if len(settingSplit) != 2 {
 					return nil, fmt.Errorf("missing =")
 				}
-				key, value := strings.TrimSpace(settingSplit[0]), strings.TrimSpace(settingSplit[1])
+				key, rawValue := strings.TrimSpace(settingSplit[0]), settingSplit[1]
+				value, err := url.QueryUnescape(rawValue)
```
Comment (MichaelS11) [inline config.go]: What do you think about moving QueryUnescape to only the values that need it?
username, password, certPath, keyPath, and caPath

File: config_test.go [hunk 6]
```diff
@@ -60,7 +60,14 @@ func TestClusterConfigToConfigString(t *testing.T) {
...
 }
 
+func sslClusterConfig(sslCfg *gocql.SslOptions) *gocql.ClusterConfig {
+	cfg := NewClusterConfig()
+	cfg.SslOpts = sslCfg
+	return cfg
+}
+
 func TestConfigStringToClusterConfig(t *testing.T) {
+
```
Comment (MichaelS11) [inline config_test.go]: Please remove space.

File: config_test.go [hunk 7]
```diff
@@ -72,8 +79,15 @@ func TestConfigStringToClusterConfig(t *testing.T) {
 		{info: "failed disableInitialHostLookup", configString: "?disableInitialHostLookup=", err: fmt.Errorf("failed for: disableInitialHostLookup = ")},
 		{info: "failed writeCoalesceWaitTime", configString: "?writeCoalesceWaitTime=", err: fmt.Errorf("failed for: writeCoalesceWaitTime = ")},
 		{info: "invalid key", configString: "?foo=bar", err: fmt.Errorf("invalid key: foo")},
+		{info: "failed enableHostVerification", configString: "?enableHostVerification=", err: fmt.Errorf("failed for: enableHostVerification = ")},
 
 		{info: "empty", configString: "", clusterConfig: NewClusterConfig()},
+		{info: "empty", configString: "?caPath=/some%20path.pem", clusterConfig: sslClusterConfig(&gocql.SslOptions{CaPath: "/some path.pem"})},
```
Comment (MichaelS11) [inline config_test.go]: Please update info on this and the following

File: config_test.go [hunk 8]
```diff
@@ -72,8 +79,15 @@ func TestConfigStringToClusterConfig(t *testing.T) {
 		{info: "failed disableInitialHostLookup", configString: "?disableInitialHostLookup=", err: fmt.Errorf("failed for: disableInitialHostLookup = ")},
 		{info: "failed writeCoalesceWaitTime", configString: "?writeCoalesceWaitTime=", err: fmt.Errorf("failed for: writeCoalesceWaitTime = ")},
 		{info: "invalid key", configString: "?foo=bar", err: fmt.Errorf("invalid key: foo")},
+		{info: "failed enableHostVerification", configString: "?enableHostVerification=", err: fmt.Errorf("failed for: enableHostVerification = ")},
```
Comment (MichaelS11) [inline config_test.go]: Please add a ParseBool fail

File: config.go (hunk 5 above)
Comment (MichaelS11) [reply config.go]: Please make sure to add QueryUnescape fail tests for each one

File: config_test.go [hunk 9]
```diff
@@ -72,8 +79,15 @@ func TestConfigStringToClusterConfig(t *testing.T) {
...
 
 		{info: "empty", configString: "", clusterConfig: NewClusterConfig()},
+		{info: "empty", configString: "?caPath=/some%20path.pem", clusterConfig: sslClusterConfig(&gocql.SslOptions{CaPath: "/some path.pem"})},
+		{info: "empty", configString: "?certPath=/some+path.pem", clusterConfig: sslClusterConfig(&gocql.SslOptions{CertPath: "/some path.pem"})},
+		{info: "empty", configString: "?keyPath=/some path.pem", clusterConfig: sslClusterConfig(&gocql.SslOptions{KeyPath: "/some path.pem"})},
+		{info: "empty", configString: "?enableHostVerification=1", clusterConfig: sslClusterConfig(&gocql.SslOptions{EnableHostVerification: true})},
+		{info: "empty", configString: "?enableHostVerification=true", clusterConfig: sslClusterConfig(&gocql.SslOptions{EnableHostVerification: true})},
+		{info: "empty", configString: "?enableHostVerification=t", clusterConfig: sslClusterConfig(&gocql.SslOptions{EnableHostVerification: true})},
```
Comment (MichaelS11) [inline config_test.go]: Please add tests that check having more than one SslOptions option

**Review by MichaelS11 — CHANGES_REQUESTED** [review]
Please add tests for everything new and username & password with QueryEscape in TestClusterConfigToConfigString

File: config_test.go [hunk 10]
```diff
@@ -60,55 +60,114 @@ func TestClusterConfigToConfigString(t *testing.T) {
...
-	tests = append(tests, TestStringToConfigStruct{info: "Host", configString: "one", clusterConfig: NewClusterConfig()})
-	tests[len(tests)-1].clusterConfig.Hosts = []string{"one"}
-	tests = append(tests, TestStringToConfigStruct{info: "Hosts", configString: "one,two,three", clusterConfig: NewClusterConfig()})
-	tests[len(tests)-1].clusterConfig.Hosts = []string{"one", "two", "three"}
-	tests = append(tests, TestStringToConfigStruct{info: "Host & Consistency any", configString: "one?consistency=any", clusterConfig: NewClusterConfig()})
-	tests[len(tests)-1].clusterConfig.Consistency = 0
-	tests[len(tests)-1].clusterConfig.Hosts = []string{"one"}
-	tests = append(tests, TestStringToConfigStruct{info: "Hosts & Consistency one", configString: "one,two,three?consistency=one", clusterConfig: NewClusterConfig()})
-	tests[len(tests)-1].clusterConfig.Consistency = 1
-	tests[len(tests)-1].clusterConfig.Hosts = []string{"one", "two", "three"}
```
Comment (choonkeat) [inline config_test.go:110]: This section of the test had slipped my attention earlier. In 33ec84e I've reorged all test cases into the same block, segmented by comment blocks instead

File: config.go (hunk 5 above)
Comment (MichaelS11) [inline config.go]: How about moving QueryUnescape to only the values that need it?
So only for username, password, certPath, keyPath, and caPath cases.

File: config_test.go (hunk 10 above)
Comment (MichaelS11) [reply config_test.go:110]: Looking good!

**Review by MichaelS11 — CHANGES_REQUESTED** [review]
Still looking at adding to TestClusterConfigToConfigString right?

**Comment by MichaelS11** [comment]
Thank you very much, looks great! :)
